File: miscellaneous/search/20230405/binary-search.php
<?php

declare(strict_types=1);

function binarySearch(array $nums, int $target, int $start=0, ?int $end=null): int
{
    $end = $end ?: count($nums) - 1;

    if ($end < $start) return -1;

    $mid = (int) floor(($end - $start) / 2) + $start;

    if ($target < $nums[$mid]) return binarySearch($nums, $target, $start, $mid);
    if ($target > $nums[$mid]) return binarySearch($nums, $target, $mid, $end);
    if ($target === $nums[$mid]) return $mid;

    return -1;
}

File: miscellaneous/search/20230405/binary-search.go
package main

func binarySearch(values []int, target int, start int, end int) int {
	mid := start + end - start/2

	if target < values[mid] {
		return binarySearch(values, target, start, mid-1)
	}

	if target > values[mid] {
		return binarySearch(values, target, mid+1, end)
	}

	if values[mid] == target {
		return mid
	}

	return -1
}

func main() {
	if binarySearch([]int{1, 2, 3, 4, 5}, 3, 0, 4) == 2 {
		println("Pass")
	} else {
		println("Fail")
	}
}
